packages/smart-contracts/src/contracts/SwapToPay.sol

```diff
@@ -0,0 +1,259 @@
+    erc20.approve(address(paymentProxy), max);
+  }
+  
+  function approveRouterToSpend(address erc20Address) public {
+    IERC20 erc20 = IERC20(erc20Address);
+    uint256 max = 2**256 - 1;
+    erc20.approve(address(uniswapRouter), max);
+  }
+  
+  /**
+  * @notice Performs a token swap between a payment currency and a request currency, and then
+  *         calls a payment proxy to pay the request, including fees.
+  * @param _to Transfer recipient = request issuer
+  * @param _amount Amount to transfer in request currency
+  * @param _amountInMax Maximum amount allowed to spend for currency swap, in payment currency.
+            This amount should take into account the fees.
+    @param _path, path of ERC20 tokens to swap from requestedToken to spentToken. The first 
+            address of the path should be the payment currency. The last element should be the
+            request currency.
+  * @param _paymentReference Reference of the payment related
+  * @param _feeAmount Amount of the fee in request currency
+  * @param _feeAddress Where to pay the fee
+  * @param _deadline Deadline for the swap to be valid
+  *
+  * TODO test the behaviour when the deadline is passed
+  */
+  function transferMkrFromDaiWithReference(
+    address _to,
+    uint256 _amount, // requestedToken
+    uint256 _amountInMax, // spentToken
+    address[] calldata _path, // from requestedToken to spentToken
+    bytes calldata _paymentReference,
+    uint256 _feeAmount,
+    address _feeAddress,
+    uint256 _deadline
+  )
+    external
+  {
+    IERC20 spentToken = IERC20(_path[0]);
+    IERC20 requestedToken = IERC20(_path[_path.length-1]);
+    
+    uint256 requestedTotalAmount = _amount + _feeAmount;
+
+    spentToken.transferFrom(msg.sender, address(this), _amountInMax);
+
+    // Allow the router to spend all this contract's spentToken
+    if (spentToken.allowance(address(this),address(uniswapRouter)) < _amountInMax) {
+        approveRouterToSpend(address(spentToken));
+    }
+    
+    // TODO? Add require
+    uniswapRouter.swapTokensForExactTokens(
+        requestedTotalAmount,
+        _amountInMax,
+        _path,
+        address(this),
+        _deadline
+    );
+    
+    // Allow the payment network to spend all this contract's requestedToken
+    //const pnAddress = 0xFb6819d605E1Fa377a932016274cC84c9A07322f;
+    if (requestedToken.allowance(address(this),address(paymentProxy)) < requestedTotalAmount) {
+        approvePaymentProxyToSpend(address(requestedToken));
+    }
+    
+    // Pay the request and fees
+    paymentProxy.transferFromWithReferenceAndFee(
+        address(requestedToken),
+        _to,
+        _amount,
+        _paymentReference,
+        _feeAmount,
+        _feeAddress
+    );
+    
+    // Give the change back
+    spentToken.transfer(msg.sender, spentToken.balanceOf(address(this)));
+  }
+  
+
+  /*
+  * Admin functions to edit the admin, router address or proxy address
+  */
+  modifier onlyAdmin() {
```

Comment:
We should probably use the `WhitelistedRole` contract that we use at the RequestHashStorage.
`WhitelistAdminRole` is also an option, but for consistency we could use the same.
Both are from the openzeppelin-solidity package.

Reply:
I used isOnwable instead, it looks like the new WhitelistedRole